File: Android_Java/wit-example-ble5/app/src/main/java/com/wit/example/ble5/components/Bwt901bleProcessor.java
package com.wit.example.ble5.components;

import android.util.Log;

import com.wit.witsdk.sensor.modular.connector.entity.BluetoothBLEOption;
import com.wit.witsdk.sensor.modular.connector.modular.bluetooth.WitBluetoothManager;
import com.wit.witsdk.sensor.modular.connector.roles.WitCoreConnect;
import com.wit.witsdk.sensor.modular.device.DeviceModel;
import com.wit.example.ble5.data.WitSensorKey;
import com.wit.witsdk.sensor.modular.processor.interfaces.IDataProcessor;
import com.wit.witsdk.sensor.utils.DipSensorMagHelper;
import com.wit.witsdk.utils.BitConvert;
import com.wit.witsdk.utils.NumberFormat;
import com.wit.witsdk.utils.StringUtils;

/**
 * 蓝牙5.0传感器数据解析
 *
 * @Author haungyajun
 * @Date 2022/4/26 15:25
 */
public class Bwt901bleProcessor implements IDataProcessor {

    /**
     * 控制自动读取线程
     */
    private boolean readDataThreadRuning = false;

    // 设备模型
    private DeviceModel deviceModel;

    @Override
    public void OnOpen(DeviceModel deviceModel) {
        this.deviceModel = deviceModel;
        readDataThreadRuning = true;
        Thread thread = new Thread(this::readDataThread);
        thread.start();
    }

    /**
     * 发送协议数据
     *
     * @author huangyajun
     * @date 2023/2/27 19:23
     */
    private void sendProtocolData(DeviceModel deviceModel, byte[] bytes, int delay) {
        deviceModel.sendProtocolData(bytes, delay);// 磁场
        try {
            Thread.sleep(delay);
        } catch (InterruptedException e) {
            e.printStackTrace();
            System.out.println(e.getMessage());
            return;
        }
    }

    /**
     * 读取数据线程
     *
     * @author huangyajun
     * @date 2022/5/24 11:39
     */
    private void readDataThread() {
        int count = 0;
        try {
            Thread.sleep(5000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        while (readDataThreadRuning) {
            try {

                String magType = deviceModel.getDeviceData("72");// 磁场类型
                if (StringUtils.IsNullOrEmpty(magType)) {
                    // 读取72磁场类型寄存器,后面解析磁场的时候要用到
                    sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, 0x27, 0x72, 0x00}, 150);
                    sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, 0x27, 0x72, 0x00}, 150);
                }

                String reg2e = deviceModel.getDeviceData("2E");// 版本号
                String reg2f = deviceModel.getDeviceData("2F");// 版本号
                if (StringUtils.IsNullOrEmpty(reg2e) || StringUtils.IsNullOrEmpty(reg2f)) {
                    // 读版本号
                    sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, 0x27, 0x2E, 0x00}, 150);
                }

                sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, (byte) 0x27, (byte) 0x3a, (byte) 0x00}, 150);// 磁场
                sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, (byte) 0x27, (byte) 0x51, (byte) 0x00}, 150);// 四元数

                // 不需要读那么快的数据
                if (count++ % 50 == 0 || count < 5) {
                    sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, (byte) 0x27, (byte) 0x64, (byte) 0x00}, 150);// 电量
                    sendProtocolData(deviceModel, new byte[]{(byte) 0xff, (byte) 0xaa, (byte) 0x27, (byte) 0x40, (byte) 0x00}, 150);// 温度


                }

                // 读取信号
                if (count % 5 == 0) {
                    WitCoreConnect coreConnect = deviceModel.getCoreConnect();
                    BluetoothBLEOption bluetoothBLEOption = coreConnect.getConfig().getBluetoothBLEOption();
                    deviceModel.setDeviceData(WitSensorKey.Rssi, WitBluetoothManager.getRssi(bluetoothBLEOption.getMac()) + "");
                }
            } catch (Exception e) {
                e.printStackTrace();
                Log.i("", "BWT901BLECL5_0DataProcessor:自动读取数据异常");
            }
        }
    }

    /**
     * 关闭设备时
     *
     * @author huangyajun
     * @date 2022/5/24 11:42
     */
    @Override
    public void OnClose() {
        // 停止线程
        readDataThreadRuning = false;
    }

    /**
     * 解算数据
     *
     * @author huangyajun
     * @date 2022/4/27 14:39
     */
    @Override
    public void OnUpdate(DeviceModel deviceModel) {
        // 加速度
        String regAx = deviceModel.getDeviceData("61_0");
        String regAy = deviceModel.getDeviceData("61_1");
        String regAz = deviceModel.getDeviceData("61_2");
        // 角速度
        String regWx = deviceModel.getDeviceData("61_3");
        String regWy = deviceModel.getDeviceData("61_4");
        String regWz = deviceModel.getDeviceData("61_5");
        // 角度
        String regAngleX = deviceModel.getDeviceData("61_6");
        String regAngleY = deviceModel.getDeviceData("61_7");
        String regAngleZ = deviceModel.getDeviceData("61_8");

        // 四元数
        String regQ1 = deviceModel.getDeviceData("51");
        String regQ2 = deviceModel.getDeviceData("52");
        String regQ3 = deviceModel.getDeviceData("53");
        String regQ4 = deviceModel.getDeviceData("54");
        // 温度和电量
        String regTemperature = deviceModel.getDeviceData("40");
        String regPower = deviceModel.getDeviceData("64");


        // 版本号
        String reg2e = deviceModel.getDeviceData("2E");// 版本号
        String reg2f = deviceModel.getDeviceData("2F");// 版本号

        // 如果有版本号
        if (StringUtils.IsNullOrEmpty(reg2e) == false &&
                StringUtils.IsNullOrEmpty(reg2f) == false) {
            short reg2eValue = Short.parseShort(reg2e);
            short reg2fValue = Short.parseShort(reg2f);

            int tempVerSion = BitConvert.byte2Int(new byte[]{
                    BitConvert.short2byte(reg2fValue)[0],
                    BitConvert.short2byte(reg2fValue)[1],
                    BitConvert.short2byte(reg2eValue)[0],
                    BitConvert.short2byte(reg2eValue)[1]
            });

            // UInt32 tempVerSion = BitConverter.ToUInt32(Buffer, 2);
            String sbinary = Integer.toBinaryString(tempVerSion);// Convert.ToString(tempVerSion, 2);
            sbinary = StringUtils.padLeft(sbinary, 32, '0');
            if (sbinary.substring(0, 1).equals("1"))//新版本号
            {
                String tempNewVS = Integer.parseInt(sbinary.substring(2, 18), 2) + "";
                tempNewVS += "." + Integer.parseInt(sbinary.substring(19, 19 + 5), 2);
                tempNewVS += "." + Integer.parseInt(sbinary.substring(25), 2);
                //Public.Common.Version_Number = tempNewVS;
                deviceModel.setDeviceData(WitSensorKey.VersionNumber, tempNewVS);
            } else {
                int tempNewVS = BitConvert.byte2Int(new byte[]{
                        0,
                        0,
                        BitConvert.short2byte(reg2eValue)[0],
                        BitConvert.short2byte(reg2eValue)[1]
                });
                deviceModel.setDeviceData(WitSensorKey.VersionNumber, tempNewVS + "");
            }
        }

        // 加速度解算
        if (!StringUtils.IsNullOrEmpty(regAx)) {
            deviceModel.setDeviceData(WitSensorKey.AccX, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regAx) / 32768 * 16) + "");
        }
        if (!StringUtils.IsNullOrEmpty(regAy)) {
            deviceModel.setDeviceData(WitSensorKey.AccY, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regAy) / 32768 * 16));
        }
        if (!StringUtils.IsNullOrEmpty(regAz)) {
            deviceModel.setDeviceData(WitSensorKey.AccZ, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regAz) / 32768 * 16));
        }

        // 角速度解算
        if (!StringUtils.IsNullOrEmpty(regWx)) {
            deviceModel.setDeviceData(WitSensorKey.AsX, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regWx) / 32768 * 2000));
        }
        if (!StringUtils.IsNullOrEmpty(regWy)) {
            deviceModel.setDeviceData(WitSensorKey.AsY, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regWy) / 32768 * 2000));
        }
        if (!StringUtils.IsNullOrEmpty(regWz)) {
            deviceModel.setDeviceData(WitSensorKey.AsZ, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regWz) / 32768 * 2000));
        }

        // 角度
        if (!StringUtils.IsNullOrEmpty(regAngleX)) {
            deviceModel.setDeviceData(WitSensorKey.AngleX, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regAngleX) / 32768 * 180));
        }
        if (!StringUtils.IsNullOrEmpty(regAngleY)) {
            deviceModel.setDeviceData(WitSensorKey.AngleY, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regAngleY) / 32768 * 180));
        }
        if (!StringUtils.IsNullOrEmpty(regAngleZ)) {
            String anZ = NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regAngleZ) / 32768 * 180);
            deviceModel.setDeviceData(WitSensorKey.AngleZ, anZ);
        }

        // 磁场
        String regHX = deviceModel.getDeviceData("3A");
        String regHY = deviceModel.getDeviceData("3B");
        String regHZ = deviceModel.getDeviceData("3C");
        // 磁场类型
        String magType = deviceModel.getDeviceData("72");
        if (!StringUtils.IsNullOrEmpty(regHX) &&
                !StringUtils.IsNullOrEmpty(regHY) &&
                !StringUtils.IsNullOrEmpty(regHZ) &&
                !StringUtils.IsNullOrEmpty(magType)
        ) {
            short type = Short.parseShort(magType);
            // 解算数据,并且保存到设备数据里
            deviceModel.setDeviceData(WitSensorKey.HX, DipSensorMagHelper.GetMagToUt(type, Double.parseDouble(regHX)) + "");
            deviceModel.setDeviceData(WitSensorKey.HY, DipSensorMagHelper.GetMagToUt(type, Double.parseDouble(regHY)) + "");
            deviceModel.setDeviceData(WitSensorKey.HZ, DipSensorMagHelper.GetMagToUt(type, Double.parseDouble(regHZ)) + "");
        }


        // 温度
        if (!StringUtils.IsNullOrEmpty(regTemperature)) {
            deviceModel.setDeviceData(WitSensorKey.T, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regTemperature) / 100));
        }

        // 电量
        if (!StringUtils.IsNullOrEmpty(regPower)) {

            int regPowerValue = Integer.parseInt(regPower);

            float eqPercent = getEqPercent((float) (regPowerValue / 100.0));
            deviceModel.setDeviceData(WitSensorKey.ElectricQuantityPercentage, eqPercent + "");
            // 电量原始值
            deviceModel.setDeviceData(WitSensorKey.ElectricQuantity, regPowerValue + "");
        }

        // 四元数
        if (!StringUtils.IsNullOrEmpty(regQ1)) {
            deviceModel.setDeviceData(WitSensorKey.Q0, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regQ1) / 32768.0));
        }
        if (!StringUtils.IsNullOrEmpty(regQ2)) {
            deviceModel.setDeviceData(WitSensorKey.Q1, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regQ2) / 32768.0));
        }
        if (!StringUtils.IsNullOrEmpty(regQ3)) {
            deviceModel.setDeviceData(WitSensorKey.Q2, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regQ3) / 32768.0));
        }
        if (!StringUtils.IsNullOrEmpty(regQ4)) {
            deviceModel.setDeviceData(WitSensorKey.Q3, NumberFormat.formatDoubleToString("%.3f", Double.parseDouble(regQ4) / 32768.0));
        }
    }

    /**
     * 获得电流值
     *
     * @author huangyajun
     * @date 2022/8/16 18:26
     */
    public float getEqPercent(float eq) {
        float p = 0;
        if (eq > 5.50) {
            p = Interp(eq,
                    new float[]{6.5f, 6.8f, 7.35f, 7.75f, 8.5f, 8.8f},
                    new float[]{0, 10, 30, 60, 90, 100});
        } else {
            p = Interp(eq,
                    new float[]{3.4f, 3.5f, 3.68f, 3.7f, 3.73f, 3.77f, 3.79f, 3.82f, 3.87f, 3.93f, 3.96f, 3.99f},
                    new float[]{0, 5, 10, 15, 20, 30, 40, 50, 60, 75, 90, 100});
        }
        return p;
    }

    public float Interp(float a, float[] x, float[] y) {
        float v = 0;
        int L = x.length;
        if (a < x[0]) v = y[0];
        else if (a > x[L - 1]) v = y[L - 1];
        else {
            for (int i = 0; i < y.length - 1; i++) {
                if (a > x[i + 1]) continue;
                v = y[i] + (a - x[i]) / (x[i + 1] - x[i]) * (y[i + 1] - y[i]);
                break;
            }
        }
        return v;
    }

    private void ReadMagType(DeviceModel deviceModel) {
        // 读取72磁场类型寄存器,后面解析磁场的时候要用到
        deviceModel.sendProtocolData(new byte[]{(byte) 0xff, (byte) 0xaa, 0x27, 0x72, 0x00});
    }
}

File: Windows_C#/Wit.Example_BWT901BLE/ble5/Components/Bwt901bleProcessor.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using Wit.SDK.Device.Device.Device.DKey;
using Wit.SDK.Modular.Sensor.Device;
using Wit.SDK.Modular.Sensor.Modular.DataProcessor.Constant;
using Wit.SDK.Modular.Sensor.Modular.DataProcessor.Context;
using Wit.SDK.Modular.Sensor.Modular.DataProcessor.Interface;
using Wit.SDK.Modular.Sensor.Modular.DataProcessor.Utils;
using Wit.SDK.Modular.Sensor.Utils;
using Wit.SDK.Utils;

namespace Wit.SDK.Modular.Sensor.Modular.DataProcessor.Roles
{
    /// <summary>
    /// 蓝牙5.0数据处理
    /// </summary>
    public class Bwt901bleProcessor : IDataProcessor
    {

        public static string PowerPercent { get; } = "PowerPercent";

        public static string Power { get; } = "Power";

        /// <summary>
        /// 读取数据线程是否运行
        /// </summary>
        private bool ReadDataThreadRuning = false;

        /// <summary>
        /// 记录key值切换器
        /// </summary>
        private RecordKeySwitch RecordKeySwitch = new RecordKeySwitch();

        /// <summary>
        /// 数据刷新的key值
        /// </summary>
        private List<string> UpdateKeys = new List<string>() { "61_0", "61_1"};

        /// <summary>
        /// 设备模型
        /// </summary>
        public DeviceModel DeviceModel { get; private set; }

        public override void OnOpen(DeviceModel deviceModel)
        {
            this.DeviceModel = deviceModel;

            // 传入刷新数据的key值
            RecordKeySwitch.Open(deviceModel, UpdateKeys);
            //启动读取线程
            Thread th = new Thread(ReadDataThread) { IsBackground = true };
            ReadDataThreadRuning = true;
            th.Start();
        }

        public override void OnClose(DeviceModel deviceModel)
        {
            ReadDataThreadRuning = false;
            // 关闭key值切换器
            RecordKeySwitch.Close();
        }

        /// <summary>
        /// 更新
        /// </summary>
        /// <param name="baseDeviceModel"></param>
        public override void OnUpdate(DeviceModel baseDeviceModel)
        {
            //解算寄存器
            ParseRegData();
        }

        /// <summary>
        /// 读取数据线程
        /// </summary>
        private void ReadDataThread()
        {
            int count = 0;

            while (ReadDataThreadRuning)
            {
                try
                {
                    // 暂停读取
                    while (DataProcessorContext.AutoReadPause) Thread.Sleep(1000);

                    // 读磁场类型
                    ReadMagType(DeviceModel);
                    
                    // 控制读取频率
                    Thread.Sleep(DataProcessorContext.AutoReadInterval);

                    DeviceModel.ReadData(ByteArrayConvert.HexStringToByteArray("FF AA 27 3A 00"));//磁场
                    DeviceModel.ReadData(ByteArrayConvert.HexStringToByteArray("FF AA 27 51 00"));//四元数


                    // 不需要读那么快的数据
                    if (++count % 50 == 0) { 
                        DeviceModel.ReadData(ByteArrayConvert.HexStringToByteArray("FF AA 27 64 00"));//电量
                        DeviceModel.ReadData(ByteArrayConvert.HexStringToByteArray("FF AA 27 40 00"));//温度
                    }

                    DeviceModel.ReadData(ByteArrayConvert.HexStringToByteArray("FF AA 27 30 00"));//时间

                    // 读产品序列号
                    ReadSerialNumberReg(DeviceModel);
                    ReadVersionNumberReg(DeviceModel);
                }
                catch (Exception e)
                {
                    //有异常但不处理
                    Debug.WriteLine("BWT901BLECL5_0DataProcessor:自动读取数据异常");
                }
            }
        }

        /// <summary>
        /// 解算寄存器
        /// </summary>
        private void ParseRegData()
        {
            // 序列号
            short? reg7f = DeviceModel.GetDeviceData(new ShortKey("7F"));// 序列号
            short? reg80 = DeviceModel.GetDeviceData(new ShortKey("80"));// 序列号
            short? reg81 = DeviceModel.GetDeviceData(new ShortKey("81"));// 序列号
            short? reg82 = DeviceModel.GetDeviceData(new ShortKey("82"));// 序列号
            short? reg83 = DeviceModel.GetDeviceData(new ShortKey("83"));// 序列号
            short? reg84 = DeviceModel.GetDeviceData(new ShortKey("84"));// 序列号
            if (reg7f !=null &&
                reg80 !=null &&
                reg81 !=null &&
                reg82 !=null &&
                reg83 !=null &&
                reg84 != null)
            {
                var sbytes = BitConverter.GetBytes((short)reg7f)
                    .Concat(BitConverter.GetBytes((short)reg80))
                    .Concat(BitConverter.GetBytes((short)reg81))
                    .Concat(BitConverter.GetBytes((short)reg82))
                    .Concat(BitConverter.GetBytes((short)reg83))
                    .Concat(BitConverter.GetBytes((short)reg84))
                    .ToArray();
                string sn = Encoding.Default.GetString(sbytes);
                DeviceModel.SetDeviceData(WitSensorKey.SerialNumber, sn);
            }


            var reg50_0 = DeviceModel.GetDeviceData(new ShortKey("30"));// 年月
            var reg50_1 = DeviceModel.GetDeviceData(new ShortKey("31"));// 日时
            var reg50_2 = DeviceModel.GetDeviceData(new ShortKey("32"));// 分秒
            var reg50_3 = DeviceModel.GetDeviceData(new ShortKey("33"));// 毫秒
            // 如果回传了时间数据包就解算时间
            if (reg50_0 != null &&
                reg50_1 != null &&
                reg50_2 != null &&
                reg50_3 != null
                )
            {
                // 解算数据,并且保存到设备数据里
                var yy = 2000 + (byte)reg50_0;
                var MM = (byte)(((short)reg50_0) >> 8);
                var dd = (byte)reg50_1;
                var hh = (byte)((short)(reg50_1) >> 8);
                var mm = (byte)reg50_2;
                var ss = (byte)((short)(reg50_2) >> 8);
                var ms = ((short)reg50_3).ToString("000");


                DeviceModel.SetDeviceData(WitSensorKey.ChipTime, $"{yy}-{MM}-{dd} {hh}:{mm}:{ss}.{ms}");
            }


            // 版本号
            var reg2e = DeviceModel.GetDeviceData(new ShortKey("2E"));// 版本号
            var reg2f = DeviceModel.GetDeviceData(new ShortKey("2F"));// 版本号

            // 如果有版本号
            if (reg2e != null &&
                reg2f != null)   
            {
                var reg2eValue = (ushort)reg2e;
                var vbytes = BitConverter.GetBytes((ushort)reg2e).Concat(BitConverter.GetBytes((ushort)reg2f)).ToArray();
                UInt32 tempVerSion = BitConverter.ToUInt32(vbytes, 0);
                string sbinary = Convert.ToString(tempVerSion, 2);
                sbinary = ("").PadLeft((32 - sbinary.Length), '0') + sbinary;
                if (sbinary.StartsWith("1"))//新版本号
                {
                    string tempNewVS = Convert.ToUInt32(sbinary.Substring(4 - 3, 14 + 3), 2).ToString();
                    tempNewVS += "." + Convert.ToUInt32(sbinary.Substring(18, 6), 2);
                    tempNewVS += "." + Convert.ToUInt32(sbinary.Substring(24), 2);
                    DeviceModel.SetDeviceData(WitSensorKey.VersionNumber, tempNewVS);
                }
                else
                {
                    DeviceModel.SetDeviceData(WitSensorKey.VersionNumber, reg2eValue.ToString());
                }
            }

            //加速度
            var regAx = DeviceModel.GetDeviceData(new ShortKey("61_0"));
            var regAy = DeviceModel.GetDeviceData(new ShortKey("61_1"));
            var regAz = DeviceModel.GetDeviceData(new ShortKey("61_2"));
            // 角速度
            var regWx = DeviceModel.GetDeviceData(new ShortKey("61_3"));
            var regWy = DeviceModel.GetDeviceData(new ShortKey("61_4"));
            var regWz = DeviceModel.GetDeviceData(new ShortKey("61_5"));
            // 角度
            var regAngleX = DeviceModel.GetDeviceData(new ShortKey("61_6"));
            var regAngleY = DeviceModel.GetDeviceData(new ShortKey("61_7"));
            var regAngleZ = DeviceModel.GetDeviceData(new ShortKey("61_8"));
            // 四元数
            var regQ1 = DeviceModel.GetDeviceData(new ShortKey("51"));
            var regQ2 = DeviceModel.GetDeviceData(new ShortKey("52"));
            var regQ3 = DeviceModel.GetDeviceData(new ShortKey("53"));
            var regQ4 = DeviceModel.GetDeviceData(new ShortKey("54"));
            // 温度和电量
            var regTemperature = DeviceModel.GetDeviceData(new ShortKey("40"));
            var regPower = DeviceModel.GetDeviceData(new ShortKey("64"));


            // 加速度解算
            if (regAx!=null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AccX, Math.Round((short)(regAx) / 32768.0 * 16, 3));
            }
            if (regAy != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AccY, Math.Round((short)(regAy) / 32768.0 * 16, 3));
            }
            if (regAz != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AccZ, Math.Round((short)(regAz) / 32768.0 * 16, 3));
            }

            // 角速度解算
            if (regWx != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AsX, Math.Round((short)(regWx) / 32768.0 * 2000, 3));
            }
            if (regWy != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AsY, Math.Round((short)(regWy) / 32768.0 * 2000, 3));
            }
            if (regWz != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AsZ, Math.Round((short)(regWz) / 32768.0 * 2000, 3));
            }

            // 角度
            if (regAngleX != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AngleX, Math.Round((short)(regAngleX) / 32768.0 * 180, 3));
            }
            if (regAngleY != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AngleY, Math.Round((short)(regAngleY) / 32768.0 * 180, 3));
            }
            if (regAngleZ != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.AngleZ, Math.Round((short)(regAngleZ) / 32768.0 * 180, 3));
            }

            // 磁场
            var regHX = DeviceModel.GetDeviceData(new ShortKey("3A"));
            var regHY = DeviceModel.GetDeviceData(new ShortKey("3B"));
            var regHZ = DeviceModel.GetDeviceData(new ShortKey("3C"));
            var magType = DeviceModel.GetDeviceData("72");// 磁场类型
            if (regHX != null &&
                regHY != null &&
                regHZ != null &&
                magType != null)
            {
                short type = short.Parse(magType);
                DeviceModel.SetDeviceData(WitSensorKey.HX, DipSensorMagHelper.GetMagToUt(type, (short)(regHX)));
                DeviceModel.SetDeviceData(WitSensorKey.HY, DipSensorMagHelper.GetMagToUt(type, (short)(regHY)));
                DeviceModel.SetDeviceData(WitSensorKey.HZ, DipSensorMagHelper.GetMagToUt(type, (short)(regHZ)));
                DeviceModel.SetDeviceData(WitSensorKey.HM, Math.Round(Math.Sqrt(Math.Pow(DipSensorMagHelper.GetMagToUt(type, (short)(regHX)), 2) + 
                                                           Math.Pow(DipSensorMagHelper.GetMagToUt(type, (short)(regHY)), 2) + 
                                                           Math.Pow(DipSensorMagHelper.GetMagToUt(type, (short)(regHZ)), 2)), 2));
            }


            // 温度
            if (regTemperature!=null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.T, Math.Round((short)(regTemperature) / 100.0, 2));
            }

            // 电量
            if (regPower != null)
            {

                int regPowerValue = (short)(regPower);

                // 计算电量百分比
                if (regPowerValue >= 396)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 100);
                }
                else if (regPowerValue >= 393 && regPowerValue < 396)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 90);
                }
                else if (regPowerValue >= 387 && regPowerValue < 393)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 75);
                }
                else if (regPowerValue >= 382 && regPowerValue < 387)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 60);
                }
                else if (regPowerValue >= 379 && regPowerValue < 382)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 50);
                }
                else if (regPowerValue >= 377 && regPowerValue < 379)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 40);
                }
                else if (regPowerValue >= 373 && regPowerValue < 377)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 30);
                }
                else if (regPowerValue >= 370 && regPowerValue < 373)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 20);
                }
                else if (regPowerValue >= 368 && regPowerValue < 370)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 15);
                }
                else if (regPowerValue >= 350 && regPowerValue < 368)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 10);
                }
                else if (regPowerValue >= 340 && regPowerValue < 350)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 5);
                }
                else if (regPowerValue < 340)
                {
                    DeviceModel.SetDeviceData(WitSensorKey.PowerPercent, 0);
                }


                // 电量原始值
                DeviceModel.SetDeviceData(Power, regPowerValue.ToString());
            }


            // 四元数
            if (regQ1 != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.Q0, Math.Round((short)regQ1 / 32768.0, 5));
            }
            if (regQ2 != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.Q1, Math.Round((short)regQ2 / 32768.0, 5));
            }
            if (regQ3 != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.Q2, Math.Round((short)regQ3 / 32768.0, 5));
            }
            if (regQ4 != null)
            {
                DeviceModel.SetDeviceData(WitSensorKey.Q3, Math.Round((short)regQ4 / 32768.0, 5));
            }
        }

        /// <summary>
        /// 读取磁场类型寄存器
        /// </summary>
        private void ReadMagType(DeviceModel deviceModel)
        {
            // 读磁场类型
            if (deviceModel.GetDeviceData("72") == null)
            {
                // 读取72磁场类型寄存器,后面解析磁场的时候要用到
                deviceModel.ReadData(new byte[] { 0xff, 0xaa, 0x27, 0x72, 0x00 });
                Thread.Sleep(20);
            }
        }

        /// <summary>
        /// 读取序列号寄存器
        /// </summary>
        /// <param name="deviceModel"></param>
        private void ReadSerialNumberReg(DeviceModel deviceModel)
        {
            // 读序列号
            if (deviceModel.GetDeviceData("7F") == null && deviceModel.GetDeviceData("82") == null)
            {
                // 读序列号
                deviceModel.ReadData(new byte[] { 0xff, 0xaa, 0x27, 0x7F, 0x00 });
                deviceModel.ReadData(new byte[] { 0xff, 0xaa, 0x27, 0x7F + 3, 0x00 });
            }
        }


        /// <summary>
        /// 读取版本号寄存器
        /// </summary>
        /// <param name="deviceModel"></param>
        private void ReadVersionNumberReg(DeviceModel deviceModel)
        {
            // 读版本号
            if (deviceModel.GetDeviceData("2E") == null)
            {
                // 读版本号
                deviceModel.ReadData(new byte[] { 0xff, 0xaa, 0x27, 0x2E, 0x00 });
                Thread.Sleep(20);
            }
        }
    }
}
